src/plugins/interactive_setup/server/plugin.ts

```diff
@@ -0,0 +1,121 @@
+  readonly #getElasticsearchConnectionStatus = () => {
+    return this.#elasticsearchConnectionStatus;
+  };
+
+  constructor(private readonly initializerContext: PluginInitializerContext) {
+    this.#logger = this.initializerContext.logger.get();
+  }
+
+  public setup(core: CorePreboot) {
+    this.#configSubscription = this.initializerContext.config
+      .create<TypeOf<typeof ConfigSchema>>()
+      .subscribe((config) => {
+        this.#config = config;
+      });
+
+    // We shouldn't activate interactive setup mode if we detect that user has already configured
+    // Elasticsearch connection manually: either if Kibana system user credentials are specified or
+    // user specified non-default host for the Elasticsearch.
+    // User can also set `interactiveSetup.forceSetup` config to `true` to force interactive setup mode.
+    const shouldActiveSetupMode =
+      this.#getConfig().forceSetup ||
+      (!core.elasticsearch.config.credentialsSpecified &&
+        core.elasticsearch.config.hosts.length === 1 &&
+        core.elasticsearch.config.hosts[0] === 'http://localhost:9200');
+    if (!shouldActiveSetupMode) {
+      this.#logger.debug(
+        'Interactive setup mode will not be activated since Elasticsearch connection is already configured.'
+      );
+      return;
+    }
+
+    let completeSetup: (result: { shouldReloadConfig: boolean }) => void;
+    core.preboot.holdSetupUntilResolved(
+      'Validating Elasticsearch connection configuration…',
+      new Promise((resolve) => {
+        completeSetup = resolve;
+      })
+    );
+
+    // If preliminary check above indicates that user didn't alter default Elasticsearch connection
+    // details, it doesn't mean Elasticsearch connection isn't configured. There is a chance that they
+    // already disabled security features in Elasticsearch and everything should work by default.
+    // We should check if we can connect to Elasticsearch with default configuration to know if we
+    // need to activate interactive setup. This check can take some time, so we should register our
+    // routes to let interactive setup UI to handle user requests until the check is complete.
+    core.elasticsearch
+      .createClient('ping')
+      .asInternalUser.ping()
+      .then(
+        (pingResponse) => {
+          if (pingResponse.body) {
+            this.#logger.debug(
+              'Kibana is already properly configured to connect to Elasticsearch. Interactive setup mode will not be activated.'
+            );
+            this.#elasticsearchConnectionStatus = ElasticsearchConnectionStatus.Configured;
+            completeSetup({ shouldReloadConfig: false });
+          } else {
+            this.#logger.debug(
+              'Kibana is not properly configured to connect to Elasticsearch. Interactive setup mode will be activated.'
+            );
+            this.#elasticsearchConnectionStatus = ElasticsearchConnectionStatus.NotConfigured;
```

Comment:
Actively checking for connectivity is different to what we discussed previously.

My understanding was that we only want to show interactive setup mode when Kibana hasn't been configured. 

This change would mean we would also show it when Kibana has been incorrectly configured or when Kibana has been correctly configured but e.g. the cluster is down or there's a temporary network issue. 

I think this could have quite a few unexpected implications so might be better to only check wether Kibana has been configured without validating connectivity.

Reply:
> This change would mean we would also show it when Kibana has been incorrectly configured or when Kibana has been correctly configured but e.g. the cluster is down or there's a temporary network issue.

If Kibana default config has changed, we won't try to do that. We check if we can connect to ES only with default configuration, that may or may not be expected to work.

> I think this could have quite a few unexpected implications so might be better to only check wether Kibana has been configured without validating connectivity.

Yeah, but the problem is that there are use cases when user explicitly disabled security or TLS in ES and we don't know until we try to connect. If we block Kibana startup in this case it can become quite troublesome for the user. Let's say they automate Kibana installations, and all of them should become online with default config.

How would you solve it?

Reply:
As discussed offline, flow will be as follows: 

![Untitled-2021-08-02-1209-2](https://user-images.githubusercontent.com/190132/127869707-2ae157f4-b058-4a78-a846-1860f017e365.png)